pkg/tasks/delete_snapshots.go

```diff
@@ -0,0 +1,203 @@
+		return fmt.Errorf("payload incorrect type for %s", config.DeleteSnapshotsTask)
+	}
+
+	logger := LogForTask(task.Id.String(), task.Typename, task.RequestID)
+	ctxWithLogger := logger.WithContext(ctx)
+
+	daoReg := dao.GetDaoRegistry(db.DB)
+	domainName, err := daoReg.Domain.Fetch(ctxWithLogger, task.OrgId)
+	if err != nil {
+		return err
+	}
+
+	rhDomainName, err := daoReg.Domain.Fetch(ctxWithLogger, config.RedHatOrg)
+	if err != nil {
+		return err
+	}
+
+	pulpClient := pulp_client.GetPulpClientWithDomain(domainName)
+
+	ds := DeleteSnapshots{
+		orgID:        task.OrgId,
+		rhDomainName: rhDomainName,
+		domainName:   domainName,
+		ctx:          ctx,
+		payload:      &opts,
+		task:         task,
+		daoReg:       daoReg,
+		pulpClient:   &pulpClient,
+	}
+	return ds.Run()
+}
+
+func (ds *DeleteSnapshots) Run() error {
+	if config.PulpConfigured() && ds.pulpClient != nil {
+		err := ds.configurePulpClient()
+		if err != nil {
+			return err
+		}
+	}
+
+	for _, snapUUID := range ds.payload.SnapshotsUUIDs {
+		templateUpdateMap := make(map[string][]models.Snapshot)
+		snap, err := ds.daoReg.Snapshot.FetchUnscoped(ds.ctx, snapUUID)
+		if err != nil {
+			return err
+		}
+		repo, err := ds.daoReg.RepositoryConfig.Fetch(ds.ctx, ds.orgID, snap.RepositoryConfigurationUUID)
+		if err != nil {
+			return err
+		}
+
+		err = ds.updateTemplatesUsingSnap(&templateUpdateMap, snap)
+		if err != nil {
+			return err
+		}
+		err = ds.daoReg.Template.DeleteTemplateSnapshot(ds.ctx, snap.UUID)
+		if err != nil {
+			return err
+		}
+
+		// #TODO: Update Latest Snapshot
+
+		if config.PulpConfigured() && ds.pulpClient != nil {
+			err = ds.deleteOrUpdatePulpContent(snap, repo, templateUpdateMap)
+			if err != nil {
+				return err
+			}
+		}
+
+		err = ds.daoReg.Snapshot.Delete(ds.ctx, snapUUID)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (ds *DeleteSnapshots) getPulpClient() pulp_client.PulpClient {
+	return *ds.pulpClient
+}
+
+func (ds *DeleteSnapshots) configurePulpClient() error {
+	repo, err := ds.daoReg.RepositoryConfig.Fetch(ds.ctx, ds.orgID, ds.payload.RepoUUID)
+	if err != nil {
+		return err
+	}
+	if repo.OrgID == config.RedHatOrg && ds.rhDomainName != "" {
+		ds.pulpClient = utils.Ptr(ds.getPulpClient().WithDomain(ds.rhDomainName))
+	} else if ds.domainName != "" {
+		ds.pulpClient = utils.Ptr(ds.getPulpClient().WithDomain(ds.domainName))
+	}
+
+	ds.pulpDistHelper = helpers.NewPulpDistributionHelper(ds.ctx, *ds.pulpClient)
+
+	return nil
+}
+
+func (ds *DeleteSnapshots) deleteOrUpdatePulpContent(snap models.Snapshot, repo api.RepositoryResponse, templateUpdateMap map[string][]models.Snapshot) error {
+	for templateUUID, snaps := range templateUpdateMap {
+		distPath, distName, err := getDistPathAndName(repo, templateUUID)
+		if err != nil {
+			return err
+		}
+
+		err = ds.pulpDistHelper.CreateOrUpdateDistribution(ds.orgID, distName, distPath, snaps[0].PublicationHref)
```

Comment:
It seems like only the first snapshot in templateUpdateMap is actually used. Can the type be `map[string]models.Snapshot`, is there a reason for it to be a list? If it's not expecting a list, I think we should be explicit about that.